# Bioinformatic-Code-
This repository holds analysis done in BIOL668 such as: Bioinformatics Code 

## Features:
* Contreras_QIIME_Tutorial (2).pdf
* Contreras_qiime2_tutorial.txt
* Contreras_BIOL668_RNAseq_tut1.Rmd
* Contreras_BIOL668_RNAseq_tut2.Rmd
* Contreras_BIOL668_RNAseq_tut3.Rmd
* algs_genome_code.txt

## Files Needed: 
* rna_counts_data.csv
* rna_map_update copy.csv
* pbmc3k_filtered_gene_bc_matrices.tar.gz
* rna_map_update copy.csv
* rna_counts_data.csv
* reads_2.fq
* reads_1.fq
* SP1.fq
* out.insub732_2_R1_fastp.fastq



## QIIME2
Follows QIIME2 tutorial and each piece of code is explained in the txt. file (Contreras_qiime2_tutorial.txt). 
Tutorial can be found here:

 https://docs.qiime2.org/2024.10/

 Example of the tutorial can be found in this file: Contreras_QIIME_Tutorial (2).pdf
 
<img width="422" alt="Screenshot 2025-05-15 at 5 13 54â€¯PM" src="https://github.com/user-attachments/assets/392b27a2-bfb8-4e62-9d0b-39eea69342d4" />

 

## RNAseq Analysis
RNAseq analysis is followed through R notebooks (


Part 1 follows a limma, Glimma and edgeR. To follow the turtorial you can find the data download (https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE63310&format=file)
Tutorial can be found here: https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html

Part 2 follows a DESeq2 tutorial to analyze gene differential expression:
https://ashleyschwartz.com/posts/2023/05/deseq2-tutorial

Part 3 follows a Single-Cell RNA sequence analysis using Seurat:
https://satijalab.org/seurat/



## Algs and Genome Project
Performs fastp, kaiju on reads_2.fq and reads_1.fq files. 
Code can be found in the algs_genome_code.txt

Example usage can be found in 

